src/metabase/lib/query.cljc

```diff
@@ -8,11 +8,68 @@
    [metabase.lib.metadata.calculation :as lib.metadata.calculation]
    [metabase.lib.normalize :as lib.normalize]
    [metabase.lib.options :as lib.options]
+   [metabase.lib.ref :as lib.ref]
    [metabase.lib.schema :as lib.schema]
    [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.lib.stage :as lib.stage]
    [metabase.lib.util :as lib.util]
+   [metabase.mbql.util.match :as mbql.match]
    [metabase.util.malli :as mu]))
 
+(defn- find-clause
+  [query stage-number target-clause]
+  (let [stage (lib.util/query-stage query stage-number)
+        [target-type _ target-id] target-clause]
+    (not-empty
+      (reduce
+        (fn [accum top-level-clause]
+          (if (mbql.match/match-one (get stage top-level-clause)
+                [target-type _ target-id] true)
+            (conj accum top-level-clause)
+            accum))
+        #{}
+        [:order-by :aggregation :breakout :filter :expressions :joins :fields]))))
+
+(defn- check-subsequent-stages!
+  "Throws if target-clause is used in a subsequent stage"
+  [previous-query query stage-number target-clause]
+  (let [[_ _ target-id] target-clause]
+    (loop [stage-number stage-number]
+      (when-let [next-stage (lib.util/next-stage-number query stage-number)]
+        (when-not (->> (lib.stage/visible-columns query next-stage)
+                       (some (fn [{:keys [lib/source id] :as _column}]
+                               (and (= :source/previous-stage source) (= target-id id)))))
+          ;; Get the ref to look for from the previous-query
+          (let [target-ref (->> (lib.stage/visible-columns previous-query next-stage)
+                                (some (fn [{:keys [lib/source id] :as column}]
+                                        (when (and (= :source/previous-stage source) (= target-id id))
+                                          (lib.ref/ref column)))))]
```

Comment:
I'm not sure if there's a better way to do this, if I have a clause in stage n, then I need to know how it was referenced in stage n+1.